src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -100,37 +87,273 @@ Status RedisStoreClient::AsyncGet(const std::string &table_name, const std::stri
+      for (auto &item : result) {
+        keys.push_back(GetKeyFromIndexTableKey(item, table_name, index_key));
+      }
+      auto on_done = [this, result, callback](const Status &status) {
+        RAY_CHECK_OK(status);
+        // Delete index keys.
+        RAY_CHECK_OK(DeleteByKeys(result, callback));
+      };
+      RAY_CHECK_OK(AsyncBatchDelete(table_name, keys, callback));
+    } else {
+      callback(status);
+    }
+  };
+
+  return scanner->ScanKeys(on_done);
+}
+
+Status RedisStoreClient::DoPut(const std::string &key, const std::string &data,
+                               const StatusCallback &callback) {
+  std::vector<std::string> args = {"SET", key, data};
+  RedisCallback write_callback = nullptr;
+  if (callback) {
+    write_callback = [callback](const std::shared_ptr<CallbackReply> &reply) {
+      auto status = reply->ReadAsStatus();
+      callback(status);
+    };
+  }
+
+  auto shard_context = redis_client_->GetShardContext(key);
+  return shard_context->RunArgvAsync(args, write_callback);
+}
+
+Status RedisStoreClient::DeleteByKeys(const std::vector<std::string> &keys,
+                                      const StatusCallback &callback) {
+  std::unordered_map<RedisContext *, std::vector<std::string>> shards;
+  for (auto &key : keys) {
+    auto shard_context = redis_client_->GetShardContext(key).get();
+    auto it = shards.find(shard_context);
+    if (it == shards.end()) {
+      shards[shard_context].push_back("DEL");
+      shards[shard_context].push_back(key);
+    } else {
+      it->second.push_back(key);
+    }
+  }
+
+  auto finished_count = std::make_shared<int>(0);
+  int size = shards.size();
+  for (auto &item : shards) {
+    auto delete_callback = [finished_count, size,
+                            callback](const std::shared_ptr<CallbackReply> &reply) {
+      ++(*finished_count);
+      if (*finished_count == size) {
+        callback(Status::OK());
+      }
+    };
+    RAY_CHECK_OK(item.first->RunArgvAsync(item.second, delete_callback));
+  }
   return Status::OK();
 }
 
+std::string RedisStoreClient::GenTableKey(const std::string &table_name,
+                                          const std::string &key) {
+  return table_name + table_separator_ + key;
+}
+
+std::string RedisStoreClient::GenTableMatchPattern(const std::string &table_name) {
+  return table_name + table_separator_ + "*";
+}
+
+std::string RedisStoreClient::GetKeyFromTableKey(const std::string &full_key,
+                                                 const std::string &table_name) {
+  auto pos = table_name.size() + table_separator_.size();
+  return full_key.substr(pos, full_key.size() - pos);
+}
+
+std::string RedisStoreClient::GenIndexTableKey(const std::string &table_name,
+                                               const std::string &key,
+                                               const std::string &index_key) {
+  return table_name + index_table_separator_ + index_key + key;
```

Comment:
1) need a separator between `index_key` and `key` as well.
2) can use the same char for `index_table_separator_` and `table_separator_`? 
e.g., `NODE:${node_id}` and `ACTOR:${job_id}:${actor_id}`.

Reply:
fixed